src/librustc_codegen_ironox/type_.rs

```diff
@@ -46,25 +47,46 @@ pub enum LLType {
         args: Vec<Type>,
         ret: Type
     },
-    /// A scalar type.
+    Array {
+        length: u64,
+    },
+    PtrTo {
+        pointee: Type,
+    },
     Scalar(ScalarType),
-    /// A placeholder for types. This will be removed later.
-    None,
+    StructType {
+        name: Option<String>,
+        members: Vec<Type>
+    },
+    Void,
 }
 
 impl CodegenCx<'ll, 'tcx> {
-    crate fn type_named_struct(&self, name: &str) -> &Type {
-        unimplemented!("type_named_struct");
+    crate fn type_named_struct(&self, name: &str) -> Type {
+        let mut borrowed_types = self.types.borrow_mut();
+        let struct_type = LLType::StructType {
+            name: Some(name.to_string()),
+            members: vec![],
+        };
+        borrowed_types.push(struct_type);
+        borrowed_types.len() - 1
     }
 
-    crate fn set_struct_body(&self, ty: &Type, els: &[&Type], packed: bool) {
-        unimplemented!("set_struct_body");
+    crate fn set_struct_body(&self, ty: Type, els: &[Type], packed: bool) {
+        let mut borrowed_types = self.types.borrow_mut();
+        if let LLType::StructType{ ref name, ref mut members } = borrowed_types[ty] {
+            *members = els.to_vec();
+            return;
+        }
+        bug!("expected LLType::StructType, found {:?}", borrowed_types[ty]);
     }
 }
 
 impl BaseTypeMethods<'tcx> for CodegenCx<'ll, 'tcx> {
     fn type_void(&self) -> Type {
-        unimplemented!("type_void");
+        let mut borrowed_types = self.types.borrow_mut();
```

Comment:
So can we end up pushing duplicates if this method is called more than once?

Reply:
Afraid so... I need to come up with a better way of passing types around.

Reply:
Could you use a `HashSet`? If you rely on the ordering of the collection, then no.

Reply:
This sounds like something that should probably be addressed in this PR if possible. Is it difficult?

Reply:
It shouldn't be *too* hard. I'll see what I can do.